ocs-ktc/src/main/kotlin/org/ostelco/prime/ocs/core/OnlineCharging.kt

```diff
@@ -96,57 +94,56 @@ object OnlineCharging : OcsAsyncRequestConsumer {
-                reservationCounter++
-                doneSignal.countDown()
-            }
+            request.msccList.forEach { mscc ->
 
-            suspend fun consumeRequestHandler(consumptionRequest: ConsumptionRequest) {
-                storage.consume(
-                        msisdn = consumptionRequest.msisdn,
-                        usedBytes = consumptionRequest.usedBytes,
-                        requestedBytes = consumptionRequest.requestedBytes) { storeResult ->
-
-                    storeResult
-                            .fold(
-                                    { storeError ->
-                                        // FixMe : should all store errors be unknown user?
-                                        logger.error(storeError.message)
-                                        responseBuilder.resultCode = ResultCode.DIAMETER_USER_UNKNOWN
-                                        doneSignal.countDown()
-                                    },
-                                    { consumptionResult ->  consumptionResultHandler(consumptionResult) }
-                            )
+                fun consumptionResultHandler(consumptionResult: ConsumptionResult) {
+                    addGrantedQuota(consumptionResult.granted, mscc, responseBuilder)
+                    addInfo(consumptionResult.balance, mscc, responseBuilder)
+                    reportAnalytics(consumptionResult, request)
+                    Notifications.lowBalanceAlert(msisdn, consumptionResult.granted, consumptionResult.balance)
+                    reservationCounter++
                 }
-            }
 
-            val requested = mscc.requested?.totalOctets ?: 0
-            if (requested > 0) {
-                consumptionPolicy.checkConsumption(
-                        msisdn = msisdn,
-                        multipleServiceCreditControl = mscc,
-                        sgsnMccMnc = request.serviceInformation.psInformation.sgsnMccMnc,
-                        apn = request.serviceInformation.psInformation.calledStationId,
-                        imsiMccMnc = request.serviceInformation.psInformation.imsiMccMnc)
-                        .bimap(
-                                { consumptionResult -> consumptionResultHandler(consumptionResult) },
-                                { consumptionRequest ->  consumeRequestHandler(consumptionRequest) }
-                        )
-            } else {
-                doneSignal.countDown()
+                suspend fun consumeRequestHandler(consumptionRequest: ConsumptionRequest) {
+                    async {
+                        storage.consume(
+                                msisdn = consumptionRequest.msisdn,
+                                usedBytes = consumptionRequest.usedBytes,
+                                requestedBytes = consumptionRequest.requestedBytes) { storeResult ->
```

Comment:
I don't think that the coroutine scope will wait til the last argument to `storage.consume` - the lambda with storeResult arg is called back.

Reply:
correct. I have seen it fail in AT. So this is not working

Reply:
Now I understand what you are talking about

Reply:
`storage.consume()` calls it callback in a different `CoroutineScope`. A new one is created when we call `writeSuspended()`